### PR TITLE
Generalize view handling with offsets and gaps

### LINKED ISSUE #1548: Generalize storage offset support for view operations
### What feature would you like to be added?

Initial storage-offset support is being introduced in #1408, but the current implementation only covers a limited set of cases:

- Size-1 dimensions that include offsets
- Contiguous base tensor layouts (for example, generic stick format)
- torch.split when splitting only along non-stick dimensions
- Following pointwise operations such as torch.clone and torch.exp

This issue tracks the work needed to generalize storage offset support to cover additional scenarios, including:

- Non-contiguous base tensor layouts (for example, produced by earlier ops such as torch.transpose)
- Additional view operations such as slicing and torch.chunk
- Splitting stick dimensions when the split occurs on a stick-boundary alignment
- Following reduction operations, where iteration-space definitions differ from pointwise ops

**Out of scope**
- Eager mode support
  * #894
- Splitting stick dimensions at offsets that do not align with stick boundaries (this requires realization)
  * #1464
- Integration with data-copy operations (ie. overwrite)
  * #1549 

### Why is this feature needed?

VIew-based ops in target modes requires this

### Anything else we need to know?

Parent issue:
- #1333

### PR DESCRIPTION
**Pull request by yoheiueda**
Thanks for sending a pull request! Please make sure you read the [contributing guidelines](https://github.com/torch-spyre/torch-spyre/blob/main/CONTRIBUTING.md).

#### What type of PR is this?

- [x] bug
- [x] feature
- [ ] documentation
- [ ] cleanup

#### What this PR does:

This PR restore elided size‑1 dimensions as explicit variables, enabling views to correctly represent offsets and gaps.

This fixes failures in complex tensor operations and now supports:

- Size-1 dimensions that include offsets
- Subsequent reduction ops

Fixes #1548 

#### Which issue(s) this PR is related to:

- #1333 
- #1548 

#### Special notes for your reviewer:

Out of scope of this PR:

- Eager mode support
- Splitting stick dimensions at offsets that do not align with stick boundaries (this requires realization)

#### Does this PR introduce a user-facing change?

No direct user-facing changes. This improves correctness and expands support for view handling.

### CODE REVIEW

**Review by moriohara — APPROVED** [review]
lgtm

**Comment by yoheiueda** [comment]
I tested this PR using `granite-3.3-8b-instruct` with the following command and was able to obtain a reasonable, correct result.

```bash
GRANITE_MODEL_PATH="/tmp/models/granite-3.3-8b-instruct"
python $SCRIPT_DIR/aiu-fms-testing-utils/scripts/inference.py --architecture hf_pretrained \
 --model_path $GRANITE_MODEL_PATH --tokenizer $GRANITE_MODEL_PATH --unfuse_weights \
 --batch_size 1 --max_new_tokens 10 --fixed_prompt_length 128 --iters 2 \
 --device_type spyre --default_dtype fp16 --timing="per-token"
```

**Output**:
```
### Instruction:
Provide a list of instructions for preparing chicken soup.

### Response:

1. Gather ingredients: 1<|end_of_text|>
```

**Components used**:
- https://github.com/foundation-model-stack/foundation-model-stack/commit/b4f36b5af526b938db506a17dcd32d468a7a91d8
- https://github.com/foundation-model-stack/aiu-fms-testing-utils/commit/95bf2792dd8399f449fb16e44515d8fa69fa46fd


**Full log**:
<details>
<summary>granite_8b.txt</summary>

```text
[ 0/ 1]: Namespace(device_type='spyre', architecture='hf_pretrained', variant=None, model_path='/tmp/models/granite-3.3-8b-instruct', model_source=None, quantization=None, int8_weight_per_channel=False, int8_activ_quant_type='per_token', int8_smoothquant=False, tokenizer='/tmp/models/granite-3.3-8b-instruct', no_use_cache=True, unfuse_weights=True, default_dtype='fp16', cast_bf16_to_fp16=False, cast_fp16_to_bf16=False, compile=False, compile_mode='default', compile_backend='inductor', compile_dynamic=False, compile_dynamic_sendnn=False, deterministic=False, distributed=False, batch_size=1, max_prompt_length=None, min_pad_length=0, fixed_prompt_length=128, max_new_tokens=10, no_early_termination=False, prompt_type='chat', prompt_path='', output_path='', timing='per-token', iters=2, verbose=0, attention_type='sdpa', prefill_chunk_size=0, stagger_load=0, stagger_update_lazyhandle=0, dist_timeout=0)
[ 0/ 1]: loading model
[ 0/ 1]: ============================================================
[ 0/ 1]: model_path=/tmp/models/granite-3.3-8b-instruct
[ 0/ 1]: linear_config={'linear_type': 'torch_linear'}
[ 0/ 1]: fused_weights=False
[ 0/ 1]: data_type=torch.float16
[ 0/ 1]: ============================================================

[ 0/ 1]: loading complete, took 18.083s
[ 0/ 1]: generating output
[ 0/ 1]: use_cache True;; do_sample False
[ 0/ 1]: ==================
/home/yohei/dt-inductor/torch-spyre/torch_spyre/_monkey_patch.py:75: UserWarning: Backend Spyre does not support int64; downcasting to int32 may change values outside the 32-bit range. You can silence this via warnings.filterwarnings(...) or spyre.set_downcast_warning(False) or TORCH_SPYRE_DOWNCAST_WARN env. variable. (Triggered internally at /home/yohei/dt-inductor/torch-spyre/torch_spyre/csrc/types_mapping.h:96.)
  return orig_to(self, *args, **kwargs)
/home/yohei/dt-inductor/test-spyre-scripts/granite/foundation-model-stack/fms/models/granite.py:280: FallbackWarning: aten.embedding.default is falling back to cpu
  x_in = self.embedding(x_in)
/home/yohei/dt-inductor/torch-spyre/torch_spyre/_inductor/customops.py:201: FallbackWarning: torch.ops.spyre.full is falling back to cpu
  warn_fallback("torch.ops.spyre.full")
/home/yohei/dt-inductor/torch-spyre/torch_spyre/_inductor/customops.py:201: FallbackWarning: torch.ops.spyre.full is falling back to cpu
  warn_fallback("torch.ops.spyre.full")
next_val=tensor([[203]]) 
/home/yohei/dt-inductor/test-spyre-scripts/granite/foundation-model-stack/fms/models/granite.py:280: FallbackWarning: aten.embedding.default is falling back to cpu
  x_in = self.embedding(x_in)
/home/yohei/dt-inductor/torch-spyre/torch_spyre/_inductor/customops.py:201: FallbackWarning: torch.ops.spyre.full is falling back to cpu
  warn_fallback("torch.ops.spyre.full")
/home/yohei/dt-inductor/torch-spyre/torch_spyre/_inductor/customops.py:201: FallbackWarning: torch.ops.spyre.full is falling back to cpu
  warn_fallback("torch.ops.spyre.full")
next_val=tensor([[203]]) 
/home/yohei/dt-inductor/test-spyre-scripts/granite/foundation-model-stack/fms/models/granite.py:280: FallbackWarning: aten.embedding.default is falling back to cpu
  x_in = self.embedding(x_in)
/home/yohei/dt-inductor/torch-spyre/torch_spyre/_inductor/customops.py:201: FallbackWarning: torch.ops.spyre.full is falling back to cpu
  warn_fallback("torch.ops.spyre.full")
next_val=tensor([[35]]) 
/home/yohei/dt-inductor/test-spyre-scripts/granite/foundation-model-stack/fms/models/granite.py:280: FallbackWarning: aten.embedding.default is falling back to cpu
  x_in = self.embedding(x_in)
/home/yohei/dt-inductor/torch-spyre/torch_spyre/_inductor/customops.py:201: FallbackWarning: torch.ops.spyre.full is falling back to cpu
  warn_fallback("torch.ops.spyre.full")
next_val=tensor([[32]]) 
/home/yohei/dt-inductor/test-spyre-scripts/granite/foundation-model-stack/fms/models/granite.py:280: FallbackWarning: aten.embedding.default is falling back to cpu
  x_in = self.embedding(x_in)
/home/yohei/dt-inductor/torch-spyre/torch_spyre/_inductor/customops.py:201: FallbackWarning: torch.ops.spyre.full is falling back to cpu
  warn_fallback("torch.ops.spyre.full")
next_val=tensor([[570]]) 
/home/yohei/dt-inductor/test-spyre-scripts/granite/foundation-model-stack/fms/models/granite.py:280: FallbackWarning: aten.embedding.default is falling back to cpu
  x_in = self.embedding(x_in)
/home/yohei/dt-inductor/torch-spyre/torch_spyre/_inductor/customops.py:201: FallbackWarning: torch.ops.spyre.full is falling back to cpu
  warn_fallback("torch.ops.spyre.full")
next_val=tensor([[3790]]) 
/home/yohei/dt-inductor/test-spyre-scripts/granite/foundation-model-stack/fms/models/granite.py:280: FallbackWarning: aten.embedding.default is falling back to cpu
  x_in = self.embedding(x_in)
/home/yohei/dt-inductor/torch-spyre/torch_spyre/_inductor/customops.py:201: FallbackWarning: torch.ops.spyre.full is falling back to cpu
  warn_fallback("torch.ops.spyre.full")
next_val=tensor([[42568]]) 
/home/yohei/dt-inductor/test-spyre-scripts/granite/foundation-model-stack/fms/models/granite.py:280: FallbackWarning: aten.embedding.default is falling back to cpu
  x_in = self.embedding(x_in)
/home/yohei/dt-inductor/torch-spyre/torch_spyre/_inductor/customops.py:201: FallbackWarning: torch.ops.spyre.full is falling back to cpu
  warn_fallback("torch.ops.spyre.full")
next_val=tensor([[44]]) 
/home/yohei/dt-inductor/test-spyre-scripts/granite/foundation-model-stack/fms/models/granite.py:280: FallbackWarning: aten.embedding.default is falling back to cpu
  x_in = self.embedding(x_in)
/home/yohei/dt-inductor/torch-spyre/torch_spyre/_inductor/customops.py:201: FallbackWarning: torch.ops.spyre.full is falling back to cpu
  warn_fallback("torch.ops.spyre.full")
next_val=tensor([[225]]) 
/home/yohei/dt-inductor/test-spyre-scripts/granite/foundation-model-stack/fms/models/granite.py:280: FallbackWarning: aten.embedding.default is falling back to cpu
  x_in = self.embedding(x_in)
/home/yohei/dt-inductor/torch-spyre/torch_spyre/_inductor/customops.py:201: FallbackWarning: torch.ops.spyre.full is falling back to cpu
  warn_fallback("torch.ops.spyre.full")
next_val=tensor([[35]]) 
[ 0/ 1]: First-token latency: 26161.074 ms
[ 0/ 1]: Average next-token latency (including first token): 18291.956 ms
[ 0/ 1]: Average next-token latency: 17417.610 ms
[ 0/ 1]: Max next-token latency: 19101.863 ms (token #2)
[ 0/ 1]: Min next-token latency: 17136.523 ms (token #7)
[ 0/ 1]: Std deviation of next-token latencies: 598.244 ms
[ 0/ 1]: Per-token timing information: 26161.074, 19101.863, 17226.153, 17176.650, 17263.285, 17177.678, 17136.523, 17174.489, 17336.268, 17165.578 ms
[ 0/ 1]: Below is an instruction that describes a task. Write a response that appropriately completes the request.

### Instruction:
Provide a list of instructions for preparing chicken soup.

### Response:

1. Gather ingredients: 1<|end_of_text|>

[ 0/ 1]: use_cache True;; do_sample False
[ 0/ 1]: ==================
/home/yohei/dt-inductor/test-spyre-scripts/granite/foundation-model-stack/fms/models/granite.py:280: FallbackWarning: aten.embedding.default is falling back to cpu
  x_in = self.embedding(x_in)
next_val=tensor([[203]]) 
next_val=tensor([[203]]) 
next_val=tensor([[35]]) 
next_val=tensor([[32]]) 
next_val=tensor([[570]]) 
next_val=tensor([[3790]]) 
next_val=tensor([[42568]]) 
next_val=tensor([[44]]) 
next_val=tensor([[225]]) 
next_val=tensor([[35]]) 
[ 0/ 1]: First-token latency: 1183.188 ms
[ 0/ 1]: Average next-token latency (including first token): 1185.739 ms
[ 0/ 1]: Average next-token latency: 1186.023 ms
[ 0/ 1]: Max next-token latency: 1224.348 ms (token #5)
[ 0/ 1]: Min next-token latency: 1147.786 ms (token #6)
[ 0/ 1]: Std deviation of next-token latencies: 29.957 ms
[ 0/ 1]: Per-token timing information: 1183.188, 1202.169, 1220.089, 1222.739, 1224.348, 1147.786, 1155.095, 1172.432, 1177.664, 1151.886 ms
[ 0/ 1]: Below is an instruction that describes a task. Write a response that appropriately completes the request.

### Instruction:
Provide a list of instructions for preparing chicken soup.

### Response:

1. Gather ingredients: 1<|end_of_text|>
```
</details>

**Review by tardieu — APPROVED** [review]
lgtm. Very nice to see the logic is much simpler now!

I wonder whether we need to/should introduce so many variables though. For instance, for `test_slice_add_2d0s1` we introduce 2 separate variables for the logically common sliced dimension in `x` and `x.clone()`. A single shared variable should be enough. We may run out of op dimensions using the current approach. This can be deferred to a separate PR (assuming this makes sense).

**Comment by tardieu** [comment]
Concretely, if we need two dim variables for `tensor0` and one for `tensor1`, the PR introduces `z0`, `z1`, `z2`. I suspect we can reuse variables across tensors, ie, introduce `z0` and `z1` for `tensor0` and then reuse `z0` for `tensor1` hence reducing the number of op dimensions.

**Comment by yoheiueda** [comment]
That makes sense. I'll follow up with a separate PR to reuse variables.
